client/question/services/FeedbackGeneratorService.js

```diff
@@ -24,12 +24,60 @@ tie.factory('FeedbackGeneratorService', [
   'RUNTIME_ERROR_FEEDBACK_MESSAGES', 'WRONG_LANGUAGE_ERRORS', 'LANGUAGE_PYTHON',
   'CLASS_NAME_AUXILIARY_CODE', 'CLASS_NAME_SYSTEM_CODE', 'PARAGRAPH_TYPE_TEXT',
   'PARAGRAPH_TYPE_CODE', 'PARAGRAPH_TYPE_SYNTAX_ERROR',
+  'PYTHON_PRIMER_BUTTON_NAME', 'EXCLUDE_LANGUAGE_UNFAMILARITY_ERROR',
+  'EXCLUDE_CONSECUTIVE_SAME_RUNTIME_ERROR', 'UNFAMILIARITY_THRESHOLD',
   function(
     FeedbackObjectFactory, TranscriptService, ReinforcementGeneratorService,
     CODE_EXECUTION_TIMEOUT_SECONDS, SUPPORTED_PYTHON_LIBS,
     RUNTIME_ERROR_FEEDBACK_MESSAGES, WRONG_LANGUAGE_ERRORS, LANGUAGE_PYTHON,
     CLASS_NAME_AUXILIARY_CODE, CLASS_NAME_SYSTEM_CODE, PARAGRAPH_TYPE_TEXT,
-    PARAGRAPH_TYPE_CODE, PARAGRAPH_TYPE_SYNTAX_ERROR) {
+    PARAGRAPH_TYPE_CODE, PARAGRAPH_TYPE_SYNTAX_ERROR,
+    PYTHON_PRIMER_BUTTON_NAME, EXCLUDE_LANGUAGE_UNFAMILARITY_ERROR,
+    EXCLUDE_CONSECUTIVE_SAME_RUNTIME_ERROR, UNFAMILIARITY_THRESHOLD) {
+
+    /**
+     * Counter to keep track of language unfamiliarity errors which include
+     * syntax errors and wrong language errors.
+     *
+     * @type {number}
+     */
+    var consecutiveLanguageUnfamiliarityCounter = 0;
+
+    /**
+     * Counter to keep track of consecutive same errors.
+     *
+     * @type {number}
+     */
+    var consecutiveSameRuntimeErrorCounter = 0;
+
+    /**
+     * Variable to store the error string immediately before the current error.
+     * Will be used to see if the user is receiving the same exact error
+     * consecutively, a possible indication of language unfamiliarity.
+     *
+     * @type {string}
+     */
+    var previousErrorString = '';
+
+    /**
+     * Resets all but one specific counter or all counters, depending on
+     * the parameter.
+     *
+     * @param {string} currentCounter
+     * @private
+     */
+    var _resetCounters = function(currentCounter) {
```

Comment:
Actually, on reading more of the file I think the design here isn't quite right. We're sort of spreading the info about which counter to increment all over the place (since the caller takes care of incrementing one counter and this one takes care of resetting all the others). That's a bit confusing.

Can we instead have a helper function along the lines of:

    updateCounters(countersToIncrement)

that increments each counter in the given list, and resets all other counters to 0? That way this function can handles all counter management and you aren't modifying the counters outside of this function, which would be cleaner.